utils/selection.py

```diff
@@ -0,0 +1,588 @@
+                return False
+
+        return text_check
+
+    # Send initial message with buttons
+    page = 0
+    pages = paginate(choices, 10)
+    embed = create_embed(page)
+    view = StatelessSelectionView(choices, page, key, query or "")
+
+    if pm:
+        embed.add_field(name="Instructions", value=f"Click buttons or type in {ctx.channel.mention}", inline=False)
+        select_msg = await ctx.author.send(embed=embed, view=view)
+    else:
+        select_msg = await ctx.send(embed=embed, view=view)
+
+    # Dual input handling loop - buttons OR text
+    text_check = create_text_check()
+    current_timeout = timeout
+
+    while True:
+        try:
+            # Wait for either interaction or text message
+            done, pending = await asyncio.wait(
+                [
+                    asyncio.create_task(
+                        ctx.bot.wait_for(
+                            "interaction",
+                            check=lambda i: i.message and i.message.id == select_msg.id and i.user.id == ctx.author.id,
+                        )
+                    ),
+                    asyncio.create_task(ctx.bot.wait_for("message", check=text_check)),
+                ],
+                return_when=asyncio.FIRST_COMPLETED,
+                timeout=current_timeout,
+            )
+
+            # Cancel any pending tasks
+            for task in pending:
+                task.cancel()
+
+            if not done:  # Timeout occurred
+                break
+
+            result = done.pop().result()
+
+            if isinstance(result, disnake.Interaction):
+                # Handle button interaction
+                await result.response.defer()
+                action = result.data.custom_id
+
+                if action == "cancel":
+                    break
+                elif action == "next":
+                    page = min(page + 1, len(pages) - 1)
+                elif action == "prev":
+                    page = max(page - 1, 0)
+                elif action.startswith("select_"):
+                    # Selection made via button
+                    try:
+                        choice_idx = int(action.split("_")[1]) - 1
+                        if 0 <= choice_idx < len(choices):
+                            selected_choice = choices[choice_idx]
+
+                            # Create expired view before deletion/return
+                            try:
+                                if not delete or pm:
+                                    expired_view = StatelessSelectionView(choices, page, key, query or "", expired=True)
+                                    await select_msg.edit(view=expired_view)
+                            except disnake.HTTPException:
+                                pass
+
+                            if delete and not pm:
+                                await select_msg.delete()
+                            return selected_choice
+                    except (ValueError, IndexError):
```

Comment:
Does this silently error on invalid selection? Maybe some sort of user feedback could be viable here.